lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1595,6 +1651,195 @@ class BuildBBRegs : public FuncOpPartialLoweringPattern {
+        // register as the appropriate pipeline register.
+        Value stageResult = stage.getResult(i);
+        bool isIterArg = false;
+        for (auto &use : stageResult.getUses()) {
+          if (auto term =
+                  dyn_cast<staticlogic::PipelineTerminatorOp>(use.getOwner())) {
+            if (use.getOperandNumber() < term.iter_args().size()) {
+              WhileOpInterface whileOp(stage->getParentOp());
+              auto reg = getComponentState().getWhileIterReg(
+                  whileOp, use.getOperandNumber());
+              getComponentState().addPipelineReg(stage, reg, i);
+              isIterArg = true;
+            }
+          }
+        }
+        if (isIterArg)
+          continue;
+
+        // Create a register for passing this result to later stages.
+        Value value = operand.get();
+        Type resultType = value.getType();
+        assert(resultType.isa<IntegerType>() &&
+               "unsupported pipeline result type");
+        auto name = SmallString<20>("stage_");
+        name += std::to_string(stage.getStageNumber());
+        name += "_register_";
+        name += std::to_string(i);
+        unsigned width = resultType.getIntOrFloatBitWidth();
+        auto reg = createReg(getComponentState(), rewriter, value.getLoc(),
+                             name, width);
+        getComponentState().addPipelineReg(stage, reg, i);
+
+        // Note that we do not use replace all uses with here as in BuildBBRegs.
+        // Instead, we wait until after BuildOpGroups, and replace all uses
+        // inside BuildPipelineGroups, once the pipeline register created here
+        // has been assigned to.
+      }
+    });
+    return success();
+  }
+};
+
+/// Builds groups for assigning registers for pipeline stages.
+class BuildPipelineGroups : public FuncOpPartialLoweringPattern {
+  using FuncOpPartialLoweringPattern::FuncOpPartialLoweringPattern;
+
+  LogicalResult
+  PartiallyLowerFuncToComp(mlir::FuncOp funcOp,
+                           PatternRewriter &rewriter) const override {
+    for (auto pipeline : funcOp.getOps<staticlogic::PipelineWhileOp>())
+      for (auto stage :
+           pipeline.getStagesBlock().getOps<staticlogic::PipelineStageOp>())
+        if (failed(buildStageGroups(pipeline, stage, rewriter)))
+          return failure();
+
+    return success();
+  }
+
+  LogicalResult buildStageGroups(staticlogic::PipelineWhileOp whileOp,
+                                 staticlogic::PipelineStageOp stage,
+                                 PatternRewriter &rewriter) const {
+    // Collect pipeline registers for stage.
+    auto pipelineRegisters = getComponentState().getPipelineRegs(stage);
+
+    // Collect group names for the prologue or epilogue.
+    SmallVector<StringAttr> prologueGroups;
+    SmallVector<StringAttr> epilogueGroups;
+
+    // Get the number of pipeline stages, excluding the termina
+    size_t numStages = whileOp.getStagesBlock().getOperations().size() - 1;
```

Comment:
assert here to ensure the size isn't 0 for `whileOp.getStagesBlock().getOperations().size()` ?

Reply:
This is guarenteed by the verifier. It's a bit of a misnomer, but the stages block contains 1 or more stages, followed by one terminator op, which is different from stages.

Reply:
I will fix up the comment here to indicate this, looks like that got cut off.